File: 240126/정수 사각형 최댓값의 최소/minimax-path-in-square.java
import java.util.*;

public class Main {
    public static int n;

    public static int[][] board;
    public static int[][] dp;

    public static int[] dr = new int[]{-1, 0};
    public static int[] dc = new int[]{0, -1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        board = new int[n][n];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                board[i][j] = sc.nextInt();
            }
        }

        dp = new int[n][n];
        System.out.println(dpf(n - 1, n - 1));
    }
    public static boolean inRange(int r, int c){
        return 0 <= r && r < n && 0 <= c && c < n;
    }
    //  dpf(r, c) = 출발점(0, 0)에서 도착점(r, c) 거쳐간 숫자의 최대값 중 최소값!
    // dpf(r, c) = max(dpf(r, c), min(dpf(r - 1, c), board[r][c]))
    public static int dpf(int r, int c) {
        if (dp[r][c] != 0) {
            return dp[r][c];
        }

        boolean isStartPoint = true;
        
        int ret = 1000001;
        for (int d = 0; d < 2; d++) {
            int pre_r = r + dr[d];
            int pre_c = c + dc[d];
            if (inRange(pre_r, pre_c)) {
                ret = Math.min(ret, Math.max(dpf(pre_r, pre_c), board[r][c]));
                isStartPoint = false;
            }
        }

        if (isStartPoint) {
            return board[r][c];
        }

        dp[r][c] = ret;

        return ret;
    }
}

File: 240217/정수 사각형 최댓값의 최소/minimax-path-in-square.py
inf = int(1e9)

n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]
dp = [[-1 for _ in range(n)] for _ in range(n)]

dr = [1, 0]
dc = [0, 1]

# 최댓값의 최소값
# dpf(r, c) = min(ret, max(dpf(r - 1, c), board[r][c]))
# dpf(r, c) = min(ret, max(dpf(r, c - 1), board[r][c]))

def in_range(r, c):
    return 0 <= r < n and 0 <= c < n


def dpf(r, c):
    if dp[r][c] != -1:
        return dp[r][c]

    if r == 0 and c == 0:
        return board[r][c]
    
    ret = inf
    if in_range(r - 1, c):
        ret = min(ret, max(dpf(r - 1, c), board[r][c]))
    
    if in_range(r, c - 1):
        ret = min(ret, max(dpf(r, c - 1), board[r][c]))

    dp[r][c] = ret

    return ret

print(dpf(n - 1, n - 1))